xwiki-platform-core/xwiki-platform-extension/xwiki-platform-extension-index/src/main/java/org/xwiki/extension/index/internal/listener/ExtensionIndexApplicationStartedListener.java

```diff
@@ -0,0 +1,121 @@
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public
+ * License along with this software; if not, write to the Free
+ * Software Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA
+ * 02110-1301 USA, or see the FSF site: http://www.fsf.org.
+ */
+package org.xwiki.extension.index.internal.listener;
+
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.List;
+
+import javax.inject.Inject;
+import javax.inject.Named;
+import javax.inject.Singleton;
+
+import org.apache.solr.client.solrj.SolrClient;
+import org.apache.solr.client.solrj.SolrQuery;
+import org.apache.solr.client.solrj.SolrServerException;
+import org.apache.solr.client.solrj.response.QueryResponse;
+import org.apache.solr.common.SolrDocument;
+import org.xwiki.component.annotation.Component;
+import org.xwiki.extension.ExtensionId;
+import org.xwiki.extension.index.internal.ExtensionIndexSolrCoreInitializer;
+import org.xwiki.extension.index.internal.ExtensionIndexSolrUtil;
+import org.xwiki.extension.repository.CoreExtensionRepository;
+import org.xwiki.extension.repository.LocalExtensionRepository;
+import org.xwiki.observation.EventListener;
+import org.xwiki.observation.event.ApplicationStartedEvent;
+import org.xwiki.observation.event.Event;
+import org.xwiki.search.solr.Solr;
+import org.xwiki.search.solr.SolrException;
+import org.xwiki.search.solr.SolrUtils;
+
+/**
+ * Cleanup the extension index of entries related to artifacts that are not available anymore (e.g., upgrade or manual
+ * manipulation of the directories storing jar files).
+ *
+ * @version $Id$
+ * @since 15.9RC1
+ * @since 15.5.3
+ */
+@Component
+@Singleton
+@Named("ExtensionIndexApplicationStartedListener")
+public class ExtensionIndexApplicationStartedListener implements EventListener
+{
+    @Inject
+    private CoreExtensionRepository coreExtensionRepository;
+
+    @Inject
+    private LocalExtensionRepository localExtensionRepository;
+
+    @Inject
+    private Solr solr;
+
+    @Inject
+    private SolrUtils solrUtils;
+
+    @Inject
+    private ExtensionIndexSolrUtil extensionIndexSolrUtil;
+
+    @Override
+    public String getName()
+    {
+        return "ExtensionIndexApplicationStartedListener";
+    }
+
+    @Override
+    public List<Event> getEvents()
+    {
+        return List.of(new ApplicationStartedEvent());
+    }
+
+    @Override
+    public void onEvent(Event event, Object source, Object data)
+    {
+        try {
+            SolrClient client = this.solr.getClient(ExtensionIndexSolrCoreInitializer.NAME);
+            SolrQuery solrQuery = new SolrQuery();
+            int batchSize = 1000;
+            QueryResponse search = client.query(solrQuery.setRows(batchSize).setStart(0));
```

Comment:
You can tell Solr to return you only the fields you need.

Reply:
done